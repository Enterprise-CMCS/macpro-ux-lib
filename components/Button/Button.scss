@use "../../assets/theme/macbis-colors" as c;

.usa-button:disabled {
  cursor: default;
}

.cms-primary-background {
  background-color: c.$primary;
}

.cms-primary-text {
  color: c.$primary;
  box-shadow: 0 0 0 2px c.$primary;
}

.usa-button--secondary-outline {
  color: white;
  box-shadow: 0 0 0 2px white;
  background-color: transparent;

  &:hover {
    background-color: transparent;
    color: c.$gray-lightest;
  }

  &:active {
    background-color: #ffffff;
  }

  &:disabled,
  &:hover:disabled {
    background-color: transparent;
    color: c.$gray-light;
    box-shadow: 0 0 0 2px c.$gray-light;
  }
}

.usa-button--inverse {
  background-color: c.$gray-lighter;
  box-shadow: none;
  color: black;

  &:hover {
    background-color: c.$gray-lightest;
    color: black;
  }

  &:active {
    background-color: #ffffff;
  }

  &:disabled,
  &:hover:disabled {
    background-color: #5b616a;
    color: black;
  }
}
