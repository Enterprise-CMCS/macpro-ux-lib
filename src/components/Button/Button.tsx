import React from "react";
import { Icon, IconChoice } from "../Icon/Icon";

type IntrinsicElements = JSX.IntrinsicElements["button"];

interface Props extends IntrinsicElements {
  buttonText: string;
  buttonVariation?: ButtonVariation;
  disabled?: boolean;
  iconName?: IconChoice;
  shiftIconLeft?: boolean;
  ariaLabel?: string;
  largeButton?: boolean;
}

/**
 * Button Component
 * @param {string}      buttonText            Renders the text contained in the button.
 * @param {string}      [buttonVariation]     Renders the style of the button.
 * @param {boolean}     [disabled]            Determines whether or not a button is enabled.
 * @param {IconChoice}  [iconName]            Handles the icon to render in the button.
 * @param {boolean}     [shiftIconLeft]       Determines if a rendered icon should render on the left side.
 * @param {string}      [ariaLabel]           Overwrites the aria label of the button element that will be read on screen readers.
 * @param {boolean}     [largeButton]         Determines whether a large button is required.
 * @param {string}      [button]              Type of specified button.
 * @param {Event}       [onClick]             Handles its behavior when the button is clicked.
 * @param {string}      [target]              Specifies a name or a keyword that indicates where to display the response that is received after clicking the button.
 */

type ButtonVariation =
  | "primary"
  | "secondary"
  | "error"
  | "success"
  | "inverse"
  | "base"
  | "link"
  | "secondary-outline";

const ButtonVariationConversion: { [key: string]: string } = {
  primary: " cms-primary-background",
  secondary: "outline cms-primary-text",
  error: "secondary",
  success: " bg-green",
  inverse: "inverse",
  base: "base",
  link: "unstyled",
  "secondary-outline": "secondary-outline",
};

export const Button: React.FC<Props> = ({
  buttonVariation = "primary",
  buttonText,
  shiftIconLeft = false,
  iconName,
  ariaLabel,
  largeButton = false,
  disabled = false,
  ...rest
}) => {
  const buttonVariationType = ButtonVariationConversion[buttonVariation] ?? "";
  return (
    <button
      disabled={disabled}
      {...rest}
      aria-label={ariaLabel || `${buttonText} button`}
      className={`display-flex usa-button usa-button--${buttonVariationType} ${
        largeButton ? "usa-button--big" : ""
      }`}
    >
      {shiftIconLeft && iconName && (
        <Icon name={iconName} color="curentColor" />
      )}
      <span className={`padding-x-05${iconName ? " padding-top-05" : ""}`}>
        {buttonText}
      </span>
      {iconName && !shiftIconLeft && (
        <div>
          <Icon name={iconName} color="curentColor" />
        </div>
      )}
    </button>
  );
};
