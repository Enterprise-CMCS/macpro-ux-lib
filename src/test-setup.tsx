import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
