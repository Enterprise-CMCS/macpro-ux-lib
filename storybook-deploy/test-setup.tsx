import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
