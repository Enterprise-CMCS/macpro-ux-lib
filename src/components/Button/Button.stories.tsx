import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "./Button";

export default {
  title: "USWDS/Base/Button",
  component: Button,
  argTypes: {
    buttonVariation: {
      description: "Renders the style of the button",
    },
    buttonText: {
      description: "Renders the text contained in the button",
    },
    disabled: {
      control: "boolean",
      description: "Determines whether or not a button is disabled",
    },
    onClick: {
      description: "Handles button behavior when clicked",
      action: "clicked",
    },
    iconName: {
      description:
        "Determines which icon that needs to be rendered from: https://designsystem.digital.gov/components/icon/. Use the icon's listed name, for example: 'accessibility_new'",
    },
    shiftIconLeft: {
      description:
        "Determines if a rendered icon should render on the left side, the iconName prop needs to be filled",
    },
    largeButton: {
      description: "Determines whether a large button is required",
    },
    ariaLabel: {
      description:
        "Overwrites the aria label of the button element that will be read on screen readers",
    },
    type: {
      options: ["submit", "reset", "button"],
      control: { type: "select" },
      description: "Type of specified button",
    },
    target: {
      options: ["_blank", "_self", "_parent", "_top"],
      control: { type: "select" },
      description:
        "Specifies a name or a keyword that indicates where to display the response that is received after clicking the button",
    },
  },
  args: {
    buttonText: "Button",
    buttonVariation: "primary",
    disabled: false,
    onClick: undefined,
    iconName: undefined,
    shiftIconLeft: false,
    type: "button",
    target: "_self",
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = ({ buttonText, ...rest }) => (
  <Button buttonText={buttonText} {...rest} />
);

export const PrimaryButton = Template.bind({});
PrimaryButton.args = {
  buttonVariation: "primary",
};

export const PrimaryButtonWithIcon = Template.bind({});
PrimaryButtonWithIcon.args = { buttonVariation: "primary", iconName: "add" };
export const SecondaryButton = Template.bind({});
SecondaryButton.args = { buttonVariation: "secondary" };
export const ErrorButton = Template.bind({});
ErrorButton.args = { buttonVariation: "error" };
export const SuccessButton = Template.bind({});
SuccessButton.args = { buttonVariation: "success" };
export const InverseButton = Template.bind({});
InverseButton.args = { buttonVariation: "inverse" };
export const BaseButton = Template.bind({});
BaseButton.args = { buttonVariation: "base" };
export const LinkButton = Template.bind({});
LinkButton.args = { buttonVariation: "link" };
export const BigButton = Template.bind({});
BigButton.args = { largeButton: true };
export const OutlineButton = Template.bind({});
OutlineButton.args = { buttonVariation: "outline" };
