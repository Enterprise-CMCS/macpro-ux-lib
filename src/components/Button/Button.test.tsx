import React from "react";
import { Button } from "./Button";
import { fireEvent, screen, render } from "../../test-setup";
import userEvent from "@testing-library/user-event";

describe("Tests for the button component.", () => {
  const mockChangeFn = jest.fn();

  it("Should render and fire a callback", () => {
    render(
      <Button data-testid="button" buttonText="Button" onClick={mockChangeFn} />
    );
    const buttonComp = screen.getByTestId("button");

    fireEvent.click(buttonComp);
    expect(buttonComp).toBeInTheDocument();
    expect(buttonComp).toHaveTextContent("Button");
    expect(mockChangeFn).toHaveBeenCalled();
  });

  it("Should render a disabled button with an aria label of disabled big button", () => {
    render(
      <Button
        data-testid="button"
        disabled
        buttonText="Disabled Button"
        onClick={mockChangeFn}
        ariaLabel="disabled big button"
        largeButton
        buttonVariation="inverse"
        iconName="add"
      />
    );
    const buttonComp = screen.getByTestId("button");

    expect(buttonComp).toBeInTheDocument();
    expect(buttonComp).toHaveTextContent("Disabled Button");
    expect(buttonComp).toBeDisabled();
    expect(buttonComp).toHaveAttribute("aria-label", "disabled big button");
  });

  it("Should render a button with an aria label of shift icon left", () => {
    render(
      <Button
        data-testid="button"
        buttonText="Test Button"
        onClick={mockChangeFn}
        ariaLabel="shift icon left"
        shiftIconLeft
        buttonVariation="base"
        iconName="add"
      />
    );
    const buttonComp = screen.getByTestId("button");

    expect(buttonComp).toBeInTheDocument();
    expect(buttonComp).toHaveTextContent("Test Button");
    expect(buttonComp).toHaveClass("usa-button--base");
    expect(buttonComp).toHaveAttribute("aria-label", "shift icon left");
  });

  describe("tooltip hover states", () => {
    type PossiblePositions = "top" | "right" | "bottom" | "left";
    const positions: Array<PossiblePositions> = [
      "top",
      "right",
      "bottom",
      "left",
    ];

    for (const position of positions) {
      const tooltipText = `The tooltip is on the ${position}`;

      it(`renders the tooltip on the ${position}`, () => {
        render(
          <div data-testid="wrapper">
            <Button
              data-testid="button"
              className="test-button"
              buttonText="Test Button"
              onClick={mockChangeFn}
              ariaLabel="tooltip button"
              withTooltip
              tooltipText={tooltipText}
              tooltipPosition={position}
            />
          </div>
        );

        const wrapper = screen.getByTestId("wrapper");
        const button = wrapper.getElementsByClassName("test-button")[0];
        const tooltip = wrapper.getElementsByClassName("usa-tooltip__body")[0];
        // Initial state
        expect(tooltip).toBeVisible();
        expect(tooltip).toHaveTextContent(tooltipText);
        expect(tooltip).not.toHaveClass("is-set");
        expect(tooltip).toHaveAttribute("aria-hidden", "true");
        expect(tooltip).not.toHaveClass(`usa-tooltip__body--${position}`);

        // Hover
        userEvent.hover(button);
        expect(tooltip).toHaveClass("is-set");
        expect(tooltip).toHaveAttribute("aria-hidden", "false");
        expect(tooltip).toHaveClass(`usa-tooltip__body--${position}`);

        // Stop hovering
        userEvent.unhover(button);
        expect(tooltip).not.toHaveClass("is-set");
        expect(tooltip).toHaveAttribute("aria-hidden", "true");
        expect(tooltip).toHaveClass(`usa-tooltip__body--${position}`);
      });
    }
  });

  describe("compontent snapshots", () => {
    it("primary big disabled", () => {
      const { container } = render(
        <Button
          data-testid="button"
          buttonText="Big Inverse Button"
          onClick={mockChangeFn}
          ariaLabel="big inverse button"
          largeButton
          buttonVariation="inverse"
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("secondary button", () => {
      const { container } = render(
        <Button
          data-testid="button"
          buttonText="Submit"
          onClick={mockChangeFn}
          ariaLabel="Secondary button"
          buttonVariation="secondary"
          iconName="add"
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
