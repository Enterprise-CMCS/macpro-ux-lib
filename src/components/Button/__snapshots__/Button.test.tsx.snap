// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the button component. compontent snapshots primary big disabled 1`] = `
<div>
  <button
    aria-label="big inverse button"
    class="display-flex flex-align-center usa-button usa-button--inverse usa-button--big"
    data-testid="button"
    formtarget="_self"
  >
    Big Inverse Button
  </button>
</div>
`;

exports[`Tests for the button component. compontent snapshots secondary button 1`] = `
<div>
  <button
    aria-label="Secondary button"
    class="display-flex flex-align-center usa-button usa-button--outline cms-primary-text"
    data-testid="button"
    formtarget="_self"
  >
    Submit
    <svg
      aria-hidden="true"
      aria-label="add icon"
      class="usa-icon--size-3"
      color="curentColor"
      fill="currentColor"
      role="img"
    >
      <use
        href="[object Object]#add"
      />
    </svg>
  </button>
</div>
`;
